datafusion/common/src/hash_utils.rs

```diff
@@ -207,6 +208,37 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
     Ok(())
 }
 
+fn hash_struct_array(
+    array: &StructArray,
+
+    random_state: &RandomState,
+    hashes_buffer: &mut [u64],
+) -> Result<()> {
+    let values = array.columns();
+    let nulls = array.nulls();
+    let num_columns = values.len();
+
+    // Skip null columns
+    let valid_indices: Vec<usize> = if let Some(nulls) = nulls {
+        nulls.valid_indices().collect()
+    } else {
+        (0..num_columns).collect()
+    };
+
+    for i in valid_indices {
+        let mut values_hashes = vec![0u64; array.len()];
```

Comment:
Why do you create a new vec for each iteration?

Reply:
create each values_hashes for each `column`. And combine those hashed element in column to one

Reply:
The `values_hashes` can be created upfront, outside of the loop.

Reply:
Or we can actually just use `hashes_buffer` which already contains the zeroed items and remove this `values_hashes`.